File: impl/dart/lib/src/mrml.dart
import "dart:io";

import "_mbf_lite.dart";
import "error.dart";

class MrmlElement {
  final String name;
  final Map<String, String> attributes;
  final List<Object> children;

  const MrmlElement(this.name, {this.attributes = const {}, this.children = const []});
}

class Mrml {
  static MrmlElement parseString(String source, {bool allowExec = false}) {
    if (allowExec) {
      throw MakrellFormatException("MRML executable embeds are not implemented in the Dart track.");
    }

    final stream = MbfLiteTokenStream(tokenizeMbfLite(source));
    final element = _parseElement(stream);
    if (!stream.isAtEnd) {
      throw MakrellFormatException("MRML source must contain a single root element.");
    }
    return element;
  }

  static MrmlElement parseFile(String path, {bool allowExec = false}) {
    return parseString(File(path).readAsStringSync(), allowExec: allowExec);
  }

  static String writeString(MrmlElement value) {
    return _writeElement(value);
  }

  static MrmlElement _parseElement(MbfLiteTokenStream stream) {
    stream.expect(MbfLiteTokenKind.lBrace, "Expected '{' as MRML element start.");
    final tag = stream.consume();
    if (tag.kind != MbfLiteTokenKind.identifier && tag.kind != MbfLiteTokenKind.string) {
      throw MakrellFormatException("MRML element name must be identifier or string.");
    }

    final attributes = <String, String>{};
    if (stream.match(MbfLiteTokenKind.lBracket)) {
      while (!stream.match(MbfLiteTokenKind.rBracket)) {
        if (stream.isAtEnd) {
          throw MakrellFormatException("Unterminated MRML attribute block.");
        }
        final name = stream.consume();
        if (name.kind != MbfLiteTokenKind.identifier && name.kind != MbfLiteTokenKind.string) {
          throw MakrellFormatException("MRML attribute names must be identifiers or strings.");
        }
        stream.expect(MbfLiteTokenKind.equals, "Expected '=' in MRML attribute.");
        final value = stream.consume();
        if (value.kind != MbfLiteTokenKind.identifier && value.kind != MbfLiteTokenKind.number && value.kind != MbfLiteTokenKind.string) {
          throw MakrellFormatException("MRML attribute values must currently be scalar.");
        }
        attributes[_nameText(name)] = _textTokenValue(value);
      }
    }

    final children = <Object>[];
    while (!stream.match(MbfLiteTokenKind.rBrace)) {
      if (stream.isAtEnd) {
        throw MakrellFormatException("Unterminated MRML element.");
      }
      if (stream.peek?.kind == MbfLiteTokenKind.lBrace) {
        children.add(_parseElement(stream));
        continue;
      }
      final token = stream.consume();
      children.add(_textTokenValue(token));
    }

    return MrmlElement(_nameText(tag), attributes: attributes, children: children);
  }

  static String _nameText(MbfLiteToken token) => token.kind == MbfLiteTokenKind.string ? token.stringValue ?? "" : token.lexeme;

  static String _textTokenValue(MbfLiteToken token) {
    return switch (token.kind) {
      MbfLiteTokenKind.string => token.stringValue ?? "",
      MbfLiteTokenKind.identifier => token.lexeme,
      MbfLiteTokenKind.number => token.lexeme,
      _ => throw MakrellFormatException("Unsupported MRML content token '${token.lexeme}'."),
    };
  }

  static String _writeElement(MrmlElement element) {
    final attrs = element.attributes.entries
        .map((entry) => ' ${entry.key}="${_escapeAttribute(entry.value)}"')
        .join();
    if (element.children.isEmpty) {
      return "<${element.name}$attrs />";
    }

    final body = element.children.map((child) {
      if (child is MrmlElement) return _writeElement(child);
      return _escapeText(child.toString());
    }).join();

    return "<${element.name}$attrs>$body</${element.name}>";
  }

  static String _escapeText(String text) {
    return text
        .replaceAll("&", "&amp;")
        .replaceAll("<", "&lt;")
        .replaceAll(">", "&gt;");
  }

  static String _escapeAttribute(String text) {
    return _escapeText(text).replaceAll('"', "&quot;");
  }
}

File: impl/jvm/src/main/java/dev/makrell/formats/Mrml.java
package dev.makrell.formats;

import dev.makrell.formats.internal.MiniMbf;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Mrml {
    private Mrml() {
    }

    public static MrmlElement parseString(String source) {
        return parseString(source, false);
    }

    public static MrmlElement parseFile(String path) {
        try {
            return parseString(Files.readString(Paths.get(path), StandardCharsets.UTF_8));
        } catch (IOException ex) {
            throw new MakrellFormatException("Could not read MRML file: " + path);
        }
    }

    public static String writeString(MrmlElement value) {
        return writeElement(value);
    }

    public static MrmlElement parseString(String source, boolean allowExec) {
        if (allowExec && source.contains("{$")) {
            throw new MakrellFormatException("Executable embeds are not supported in the JVM track yet.");
        }
        List<MiniMbf.Node> roots = MiniMbf.parse(source);
        if (roots.size() != 1 || !"brace".equals(roots.get(0).kind)) {
            throw new MakrellFormatException("MRML expects exactly one root element.");
        }
        return parseElement(roots.get(0));
    }

    private static MrmlElement parseElement(MiniMbf.Node node) {
        if (!"brace".equals(node.kind) || node.children.isEmpty()) {
            throw new MakrellFormatException("Invalid MRML element.");
        }
        MiniMbf.Node head = node.children.get(0);
        if (!"scalar".equals(head.kind)) {
            throw new MakrellFormatException("MRML element name must be a scalar.");
        }
        String name = head.text;
        LinkedHashMap<String, String> attributes = new LinkedHashMap<>();
        List<Object> children = new ArrayList<>();
        int index = 1;
        if (index < node.children.size() && "square".equals(node.children.get(index).kind)) {
            parseAttributes(node.children.get(index), attributes);
            index++;
        }
        for (; index < node.children.size(); index++) {
            MiniMbf.Node child = node.children.get(index);
            if ("brace".equals(child.kind)) {
                children.add(parseElement(child));
            } else if ("scalar".equals(child.kind)) {
                children.add(child.text);
            } else {
                throw new MakrellFormatException("Unsupported MRML child node kind: " + child.kind);
            }
        }
        return new MrmlElement(name, attributes, children);
    }

    private static void parseAttributes(MiniMbf.Node node, LinkedHashMap<String, String> attributes) {
        int index = 0;
        while (index < node.children.size()) {
            MiniMbf.Node key = node.children.get(index++);
            if (index >= node.children.size() || !"scalar".equals(key.kind)) {
                throw new MakrellFormatException("Invalid MRML attribute list.");
            }
            if ("scalar".equals(node.children.get(index).kind) && "=".equals(node.children.get(index).text)) {
                index++;
            }
            if (index >= node.children.size()) {
                throw new MakrellFormatException("Missing MRML attribute value.");
            }
            MiniMbf.Node value = node.children.get(index++);
            if (!"scalar".equals(value.kind)) {
                throw new MakrellFormatException("MRML attribute values must be scalar.");
            }
            attributes.put(key.text, value.text);
        }
    }

    private static String writeElement(MrmlElement element) {
        StringBuilder builder = new StringBuilder();
        builder.append("<").append(element.getName());
        for (Map.Entry<String, String> entry : element.getAttributes().entrySet()) {
            builder.append(" ")
                .append(entry.getKey())
                .append("=\"")
                .append(escape(entry.getValue()))
                .append("\"");
        }
        List<Object> children = element.getChildren();
        if (children.isEmpty()) {
            builder.append("/>");
            return builder.toString();
        }
        builder.append(">");
        for (Object child : children) {
            if (child instanceof MrmlElement) {
                builder.append(writeElement((MrmlElement) child));
            } else {
                builder.append(escape(String.valueOf(child)));
            }
        }
        builder.append("</").append(element.getName()).append(">");
        return builder.toString();
    }

    private static String escape(String value) {
        return value
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;");
    }
}
